### PR TITLE
file parsing and memory usage optimization

### PR DESCRIPTION
**Pull request by LoganDark**
This PR reworks the file parsing and the memory usage estimates to make the library a bit more efficient in those respects and also pave the way towards more interesting things like new types of loading (mmap, quantize on load, load from PTH). It also reworks the type system a bit to simplify changing it up if the file format ever needs another update.

The computation graph creation has been separated out and memory estimates for it have been added so that the perfect amount of memory can be allocated each time. And the mysterious computation graph work tensor has been understood well enough to add an accurate upper bound.

This is because I plan to implement sequence mode and having a tight understanding and control of memory usage is crucial for implementing it properly. I don't want to be wasteful on large or small models, or undershoot and crash the application.

Also, the quantization has been optimized a bit and a cuBLAS offloading function has been added, to preserve backwards compatibility. The signature of functions with the same name cannot change, so adding a new parameter to `rwkv_init_from_file` is a bad idea.

This PR works with both the Raven family of models as well as the brand new World family which honestly makes me really excited.

Try to keep up :)

### CODE REVIEW

**Comment by saharNooby** [comment]
If you expect something to break, I strongly suggest adding new tests for new behaviors. `tiny-rwkv` can be used as a test model, it's already in the repo.

**Comment by LoganDark** [comment]
so maintaining the fstat, fseek etc. defines is getting weird. Microsoft is special and refuses to offer library functions that aren't prefixed by an underscore, Apple is special and only offers `fstato` which works with the width of `off_t` (which you can only make 64-bit globally), but they removed everything with ASi and only offer regular old `fstat` which is 64-bit now. and Linux is at least straightforward, and has about 3 different methods (with no documentation as to which one is current) to enable what they call "Large File Support".

I have the resources to look into this, I have multiple versions of Windows, multiple installations of Linux and an Intel Mac (and I can make up for the ASi Macs easily). But it's mildly infuriating. I will just put in the effort now and try to get it all in one go

**Comment by LoganDark** [comment]
I'm super proud of how smooth it was to update ggml with the new type system, definitely seems to be working as intended.

OK, so here is what appears to be working as intended in this PR :

- Initializing model and eval (only question: is total upper bound of context graph guaranteed to be covered by my calculation. I need to do some deeper math to confirm this, but I suspect it's very close)
- Model quantization (I just fixed what I think was the last remaining quantization regression which is that the whole model data type did not get updated on conversion)
- Compiling on all 3 major platforms, (Windows, Linux, macOS) with large 64-bit file support in all three

I am working on Q8_1 quantization right now (it is a somewhat special case; i hope it will be worth it)

**Comment by LoganDark** [comment]
> I am working on Q8_1 quantization right now (it is a somewhat special case; I hope it will be worth it)

So I successfully implemented a quantizer for it, but GGML actually does not support `vec_dot_q` for `Q8_1` tensors, which means trying to run inference on such a model results in a NULL pointer dereference at runtime. https://github.com/ggerganov/ggml/issues/200

Anyway, I removed the data type as it's not usable, but left the quantizer in place so that the type can be re-enabled once GGML supports the proper operators to perform inference on a Q8_1 quantized model.

**Comment by ggerganov** [comment]
`Q8_1` is not supported because it does not improve the accuracy compared to `Q8_0` in any measurable way

**Comment by LoganDark** [comment]
> `Q8_1` is not supported because it does not improve the accuracy compared to `Q8_0` in any measurable way

good to know, thanks. So there really is no reason to support it as an alternative to Q8_0?

**Comment by ggerganov** [comment]
Yup, no point. The existing `Q8_1` in the code base is actually the same as `Q8_0` but with precomputed sums in the `s` member as an optimization technique. It is used only for intermediated quantizations during inference. But it is not used to store model data

**Comment by LoganDark** [comment]
> Yup, no point. The existing `Q8_1` in the code base is actually the same as `Q8_0` but with precomputed sums in the `s` member as an optimization technique. It is used only for intermediate quantizations during inference. But it is not used to store model data

is it same to assume that any data type not implemented in `ggml_quantize_chunk` will never be usable for models?

File: rwkv.cpp [hunk 1]
```diff
@@ -413,26 +413,38 @@ void rwkv_max_impl(const int n_cols, float * dest, const float * src0, const flo
     }
 }
 
-struct ggml_tensor * rwkv_exp(ggml_context * ctx, struct ggml_tensor * x) {
+struct ggml_tensor * rwkv_op_inplace(struct ggml_context * ctx, enum ggml_op op, struct ggml_tensor * a, struct ggml_tensor * b) {
```
Comment (saharNooby) [inline rwkv.cpp:416]: >I don't think the memory savings are that worth it

I agree and would prefer to code to be simpler, at the cost of being slightly less optimal. The bottleneck is in matmul anyway, no point to optimize activation functions

**Comment by saharNooby** [comment]
@LoganDark Would be great to have summary of the changes in the PR desc, for easier review (it's okay to add it way later, when PR is out of draft status tho)

**Comment by LoganDark** [comment]
> @LoganDark Would be great to have summary of the changes in the PR desc, for easier review (it's okay to add it way later, when PR is out of draft status tho)

the most recent changes are preparing to add the seq mode. should that be in this PR or a separate one?

this PR might be ready, just have to verify the most recent commit

**Comment by saharNooby** [comment]
@LoganDark 

>should that be in this PR or a separate one?

I would very much prefer it :) Reviewing large PRs is difficult and leads to procrastination on them...

**Comment by LoganDark** [comment]
> large PRs

sorry about refactoring so much of your code each PR, it's just so filled with opportunity !! (lol)

**Comment by saharNooby** [comment]
>it's just so filled with opportunity !! (lol)

I see what you mean lol :) C/C++ is not my native language, so yeah, not the best code in the world...

**Comment by LoganDark** [comment]
looks like that is the kind of reason that I needed to test this PR (quantized models were failing to load because the memory estimate was too low)

**Comment by LoganDark** [comment]
wait I did that merge very wrongly ugh stupid cli tpo

**Comment by LoganDark** [comment]
ok it looks like the new function with gpu layers has to be renamed OR I can just add a new function that offloads specified layers onto gpu. going to do the latter

**Comment by LoganDark** [comment]
going to mark this as ready for review now, I think it is ok

**Comment by LoganDark** [comment]
it looks like this even supports loading and inference on the new world models, that is funny and cool

**Comment by LoganDark** [comment]
The original post has been edited with a description which should show up in the commit if you squash when merging

File: rwkv/rwkv_cpp_model.py [hunk 2]
```diff
@@ -32,11 +32,12 @@ def __init__(
...
         assert os.path.isfile(model_path), f'{model_path} is not a file'
         assert thread_count > 0, 'Thread count must be positive'
-        assert gpu_layers_count > 0, 'GPU layers count must be positive'
+        assert gpu_layers_count >= 0, 'GPU layers count must be >= 0'
 
         self._library = shared_library
 
-        self._ctx = self._library.rwkv_init_from_file(model_path, thread_count, gpu_layers_count)
+        self._ctx = self._library.rwkv_init_from_file(model_path, thread_count)
+        self._library.rwkv_gpu_offload_layers(self._ctx, gpu_layers_count)
```
Comment (saharNooby) [inline rwkv/rwkv_cpp_model.py:40]: We can skip the call if `gpu_layers_count == 0`

File: rwkv/rwkv_cpp_shared_library.py [hunk 3]
```diff
@@ -83,11 +86,13 @@ def rwkv_init_from_file(self, model_file_path: str, thread_count: int, gpu_layer
...
 
         ptr = self.library.rwkv_init_from_file(model_file_path.encode('utf-8'),
-                                               ctypes.c_uint32(thread_count),
-                                               ctypes.c_uint32(gpu_layers_count))
+                                               ctypes.c_uint32(thread_count))
         assert ptr is not None, 'rwkv_init_from_file failed, check stderr'
         return RWKVContext(ptr)
 
+    def rwkv_gpu_offload_layers(self, ctx: RWKVContext, gpu_layers_count: int) -> None:
+        assert self.library.rwkv_gpu_offload_layers(ctx.ptr, ctypes.c_uint32(gpu_layers_count)), 'rwkv_gpu_offload_layers failed, check stderr'
```
Comment (saharNooby) [inline rwkv/rwkv_cpp_shared_library.py:94]: Please add doc comment as in all other methods in the file (just copy-paste from header file, but in Python doc style)

File: rwkv.cpp [hunk 4]
```diff
@@ -7,24 +7,37 @@
...
+#else
+#include <sys/stat.h>
+#if !defined(__APPLE__)
+#define ftell ftello
+#define fseek fseeko
+#endif
 #endif
 
+static_assert(sizeof(stat::st_size) >= 8, "file offsets should be 64-bit");
+static_assert(sizeof(decltype(ftell(NULL))) >= 8, "file offsets should be 64-bit");
```
Comment (saharNooby) [inline rwkv.cpp:39]: What this does exactly?

File: rwkv.cpp [hunk 5]
```diff
@@ -38,41 +51,73 @@ inline enum rwkv_error_flags operator|=(enum rwkv_error_flags & a, enum rwkv_err
     return a = a | b;
 }
 
-// If the condition x is false, adds ERR_VAL to the last error, prints a message to stderr, and returns RET_VAL.
-#define RWKV_ASSERT_MSG(ERR_VAL, RET_VAL, x, ...) \
+#define RWKV_MSG(...) do { if (global_print_errors) fprintf(stderr, __VA_ARGS__); } while (0)
+#define RWKV_CTX_MSG(ctx, ...) do { if ((ctx)->print_errors) fprintf(stderr, __VA_ARGS__); } while (0)
```
Comment (saharNooby) [inline rwkv.cpp:55]:
```suggestion
#define RWKV_CTX_MSG(ctx, ...) do { if (ctx->print_errors) fprintf(stderr, __VA_ARGS__); } while (0)
```

File: rwkv.cpp [hunk 6]
```diff
@@ -84,72 +129,241 @@ inline enum rwkv_error_flags operator|=(enum rwkv_error_flags & a, enum rwkv_err
...
+    TYPE_COUNT,  /* I8    */
+    TYPE_COUNT,  /* I16   */
+    TYPE_COUNT,  /* I32   */
+    TYPE_COUNT,  /* COUNT */
+};
+
+static const char * type_to_string[TYPE_COUNT + 1] = {"float32", "float16", "Q4_0", "Q4_1", "Q4_1_O", "Q4_2", "Q4_3", "Q5_0", "Q5_1", "Q8_0", "unknown"};
+
+enum rwkv_type type_from_string(const char * str) {
+    for (int ord = 0; ord < TYPE_COUNT; ord++)
```
Comment (saharNooby) [inline rwkv.cpp:226]:
```suggestion
    for (int ord = 0; ord < TYPE_COUNT; ord++) {
```
...and below :)

File: rwkv.cpp [hunk 7]
```diff
@@ -84,72 +129,241 @@ inline enum rwkv_error_flags operator|=(enum rwkv_error_flags & a, enum rwkv_err
...
+        if (strcmp(str, type_to_string[ord]) == 0)
+            return (enum rwkv_type) ord;
+
+    return TYPE_UNKNOWN;
+}
+
+struct file_header {
```
Comment (saharNooby) [inline rwkv.cpp:233]:
```suggestion
struct rwkv_file_header {
```
Not a blocker, but I would prefer to prefix `rwkv.cpp` specific structures and functions with `rwkv_`. Just a style preference

File: rwkv.cpp [hunk 8]
```diff
@@ -84,72 +129,241 @@ inline enum rwkv_error_flags operator|=(enum rwkv_error_flags & a, enum rwkv_err
...
+};
+
+bool is_file_version_in_range(uint32_t version) {
+    return version >= RWKV_FILE_VERSION_MIN && version <= RWKV_FILE_VERSION_MAX;
+}
+
+bool fread_file_header(FILE * file, struct file_header & header, bool verify_data_type = true) {
+    RWKV_ASSERT_FALSE(RWKV_ERROR_FILE_READ, fread_data(file, sizeof(struct file_header), &header));
+    RWKV_ASSERT_FALSE(RWKV_ERROR_FILE_MAGIC, header.magic == RWKV_FILE_MAGIC);
+    RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_FILE_VERSION, is_file_version_in_range(header.version), "unsupported file version %" PRId32, header.version);
```
Comment (saharNooby) [inline rwkv.cpp:249]:
```suggestion
    RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_FILE_VERSION, is_file_version_in_range(header.version), "Unsupported file version %" PRId32, header.version);
```
Most messages start with upper-case letters, let's keep them consistent

File: rwkv.cpp [hunk 9]
```diff
@@ -265,194 +465,382 @@ struct ggml_tensor * rwkv_layer_norm(ggml_context * ctx, struct ggml_tensor * x,
...
+    /*  e2 */ ctx_size_add_tensor(ctx_size, 2, 0, GGML_TYPE_F32, n_embed);
+    /*  e2 */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_I32, ptr_nelem);
+
+    /*  xx */ ctx_size_add_tensor(ctx_size, 0, 0, GGML_TYPE_F32, n_embed);
+    /*  aa */ ctx_size_add_tensor(ctx_size, 2, 1, GGML_TYPE_F32, n_embed);
+    /*  bb */ ctx_size_add_tensor(ctx_size, 1, 1, GGML_TYPE_F32, n_embed);
+    /*  pp */ ctx_size_add_tensor(ctx_size, 0, 0, GGML_TYPE_F32, n_embed);
+
+    /* wkv */ ctx_size_add_tensor(ctx_size, 1, 0, GGML_TYPE_F32, n_embed);
+    /*   x */ ctx_size_add_tensor(ctx_size, 2, 1, GGML_TYPE_F32, n_embed);
```
Comment (saharNooby) [inline rwkv.cpp:633]: I hope this will go away when ggml provides automatic way for calculating/allocating context...

File: rwkv.cpp [hunk 10]
```diff
@@ -265,194 +465,382 @@ struct ggml_tensor * rwkv_layer_norm(ggml_context * ctx, struct ggml_tensor * x,
...
+    // e1 = torch.exp(ww - qq)
+    e1 = rwkv_exp(ctx, ggml_sub(ctx, ww, qq));
+    // e2 = torch.exp(k - qq)
+    e2 = rwkv_exp(ctx, ggml_sub(ctx, k, qq));
+
+    // state[5 * i + 1] = x0
+    // state[5 * i + 2] = e1 * aa + e2 * v
+    // state[5 * i + 3] = e1 * bb + e2
+    // state[5 * i + 4] = qq
+
```
Comment (saharNooby) [inline rwkv.cpp:694]: Why people keep adding this empty line here? 🤔 This is a second or third PR from different person I've seen. This block of comments directly relates to code below, and is similar to comments above -- which do not have an empty line after them.

File: rwkv.cpp [hunk 11]
```diff
@@ -466,159 +854,131 @@ struct rwkv_ggml_guard {
...
-        RWKV_ASSERT_NULL(RWKV_ERROR_MODEL_PARAMS, set_block_parameter(&parameters, i, "ffn.key.weight", &layer->ffn_key));
-        RWKV_ASSERT_NULL(RWKV_ERROR_MODEL_PARAMS, set_block_parameter(&parameters, i, "ffn.value.weight", &layer->ffn_value));
-        RWKV_ASSERT_NULL(RWKV_ERROR_MODEL_PARAMS, set_block_parameter(&parameters, i, "ffn.receptance.weight", &layer->ffn_receptance));
-    }
+    std::unique_ptr<struct rwkv_context> rwkv_ctx(new(std::nothrow) struct rwkv_context());
+    RWKV_ASSERT_NULL_MSG(RWKV_ERROR_CTX | RWKV_ERROR_ALLOC, rwkv_ctx.get(), "failed to allocate context");
 
-    int n_gpu = 0;
-    size_t vram_total = 0;
+    // don't free ggml context
```
Comment (saharNooby) [inline rwkv.cpp:956]:
```suggestion
    // Don't free ggml context
```
... and other comments

File: rwkv.cpp [hunk 12]
```diff
@@ -629,326 +989,197 @@ struct rwkv_context * rwkv_init_from_file(const char * file_path, const uint32_t
...
-    for (size_t i = 0; i < n_layer * 5; i++) {
-        struct ggml_tensor * part = graph->state_parts[i];
-        memcpy(state_out + i * n_embed, part->data, part->ne[0] * sizeof(float));
+    for (size_t i = 0; i < header.n_layer * 5; i++) {
+        struct ggml_tensor * part = graph.output_state[i];
+        memcpy(state_out + i * header.n_embed, part->data, ggml_nbytes(part));
     }
 
-    memcpy(logits_out, graph->logits->data, graph->logits->ne[0] * sizeof(float));
+    if (logits_out)
```
Comment (saharNooby) [inline rwkv.cpp:1028]:
```suggestion
    if (logits_out) {
```

File: rwkv.cpp [hunk 13]
```diff
@@ -957,18 +1188,18 @@ const char * rwkv_get_system_info_string(void) {
...
+    s += "AVX512="    + std::to_string(ggml_cpu_has_avx512())    + " ";
+    s += "FMA="       + std::to_string(ggml_cpu_has_fma())       + " ";
+    s += "NEON="      + std::to_string(ggml_cpu_has_neon())      + " ";
+    s += "ARM_FMA="   + std::to_string(ggml_cpu_has_arm_fma())   + " ";
+    s += "F16C="      + std::to_string(ggml_cpu_has_f16c())      + " ";
+    s += "FP16_VA="   + std::to_string(ggml_cpu_has_fp16_va())   + " ";
+    s += "WASM_SIMD=" + std::to_string(ggml_cpu_has_wasm_simd()) + " ";
+    s += "BLAS="      + std::to_string(ggml_cpu_has_blas())      + " ";
+    s += "SSE3="      + std::to_string(ggml_cpu_has_sse3())      + " ";
+    s += "VSX="       + std::to_string(ggml_cpu_has_vsx());
```
Comment (saharNooby) [inline rwkv.cpp:1202]: Why these changes?

File: rwkv.cpp [hunk 14]
```diff
@@ -629,326 +989,197 @@ struct rwkv_context * rwkv_init_from_file(const char * file_path, const uint32_t
...
 
-        std::string name(key_length, 0);
-        RWKV_ASSERT_NULL_MSG(RWKV_ERROR_MODEL_PARAMS | RWKV_ERROR_KEY, fread(&name[0], key_length, 1, file_in) == 1, "Failed to read parameter key");
-
-        printf("%48s - [%5d, %5d], type = %6s ", name.data(), x, y, ggml_type_name(parameter_ggml_type));
-        total_size_orig += (size_t) (nelements * ggml_type_sizef(parameter_ggml_type));
-
-        // Quantize only 2D tensors, except embedding and head matrices.
-        // Embedding and head take not too much space, especially in bigger models;
-        // but they significantly increase perplexity when quantized.
```
Comment (saharNooby) [inline rwkv.cpp:838]: I propose to keep this comment for some additional clarity

File: rwkv.h [hunk 15]
```diff
@@ -83,8 +83,11 @@ extern "C" {
     // Returns NULL on any error. Error messages would be printed to stderr.
     // - model_file_path: path to model file in ggml format.
     // - n_threads: count of threads to use, must be positive.
-    // - n_gpu_layer: count of layers need to load to gpu (only works when cuBLAS is on)
-    RWKV_API struct rwkv_context * rwkv_init_from_file(const char * model_file_path, const uint32_t n_threads, const uint32_t n_gpu_layers);
+    RWKV_API struct rwkv_context * rwkv_init_from_file(const char * model_file_path, const uint32_t n_threads);
+
+    // Offloads specified layers of context onto GPU using cuBLAS, if it is enabled.
+    // If rwkv.cpp was compiled without cuBLAS support, this function is a no-op.
+    RWKV_API bool rwkv_gpu_offload_layers(const struct rwkv_context * ctx, const uint32_t n_gpu_layers);
```
[the remark this answers is not in the document]
Comment (LoganDark) [reply rwkv.h:90]: Then ggml will pick only one and we should pick that one: https://github.com/ggerganov/ggml/blob/4078f637b078c5099da898d139b1c450258d869d/src/ggml.c#L3933-L3937

File: rwkv.cpp (hunk 4 above)
Comment (LoganDark) [reply rwkv.cpp:39]: this asserts that it supports files over 2GB

File: rwkv.cpp (hunk 13 above)
Comment (LoganDark) [reply rwkv.cpp:1202]: The line used to be so long that it wouldn't even fit on my screen

File: rwkv/rwkv_cpp_shared_library.py (hunk 3 above)
Comment (LoganDark) [reply rwkv/rwkv_cpp_shared_library.py:94]: sorry, I was in visual studio at the time which does not support forwarding ctrl+k to `nano` for copy and paste, I will fix it in windows terminal

File: rwkv.cpp (hunk 7 above)
Comment (LoganDark) [reply rwkv.cpp:233]: sorry, I think I tried this but didn't know rwkv_ is also for internal functions and not just api functions

File: rwkv.cpp (hunk 8 above)
Comment (LoganDark) [reply rwkv.cpp:249]: I actually tried to make them all consistently lowercase but ok

File: rwkv.cpp (hunk 9 above)
Comment (LoganDark) [reply rwkv.cpp:633]: I think it's pretty but the amount of work was very high so yes I do hope that it can go away

File: rwkv.cpp (hunk 14 above)
Comment (LoganDark) [reply rwkv.cpp:838]: sorry about that, I didn't remove this on purpose, just forgot to add it back

File: rwkv.cpp [hunk 16]
```diff
@@ -629,326 +989,197 @@ struct rwkv_context * rwkv_init_from_file(const char * file_path, const uint32_t
...
+        RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_MODEL_PARAMS, fread_string(in_file, header.key_length, name), "failed to read tensor name");
 
-        if (quantize) {
-            printf("quantizing... ");
-            // For quantization
-            work.resize(nelements);
+        const char * name_str = name.c_str();
+        RWKV_MSG("%*s - [%5" PRId32 ", %5" PRId32 "], type = %6s ", (int) max_key_length, name_str, header.width, header.height, type_to_string[header.data_type]);
 
-            // This is a histogram of quantized values. If it shows single 1.0, then all 0.0, something went very wrong!
```
[the remark this answers is not in the document]
Comment (LoganDark) [reply rwkv.cpp:899]: > In general, I don't see much value in removing clarifying comments

I'm sorry, I did not do it on purpose, I forgot to add them back

**Comment by LoganDark** [comment]
> error messages must start with upper-case letter

this is not documented in the code style :( I thought it was the complete opposite

> keep still relevant comments that were in previous code instead of removing them

I forgot to add them back, I didn't mean to remove them, I am sorry

File: rwkv.cpp (hunk 4 above)
Comment (saharNooby) [reply rwkv.cpp]: I then propose adding a clarifying comment to the asserts :)

Comment (LoganDark) [reply rwkv.cpp]: I think I actually did when you made this comment, the more recent commits made the error messages a bit more descriptive

Comment (saharNooby) [reply rwkv.cpp]: Nvm, you've already changed the messages, I'm ok with them

File: rwkv.cpp [hunk 17]
```diff
@@ -875,15 +875,15 @@ struct rwkv_context * rwkv_init_from_file(const char * file_path, const uint32_t
...
     rwkv_file_guard file_guard { file };
 
     // Be very careful when changing this code. It must support files larger than 2 GB by using 64-bit functions to the get file length.
     struct stat file_stat;
-    RWKV_ASSERT_NULL_MSG(RWKV_ERROR_FILE | RWKV_ERROR_FILE_STAT, fstat(fileno(file), &file_stat) == 0, "failed to stat file %s", file_path);
+    RWKV_ASSERT_NULL_MSG(RWKV_ERROR_FILE | RWKV_ERROR_FILE_STAT, fstat(fileno(file), &file_stat) == 0, "Failed to stat file %s", file_path);
 
     struct rwkv_file_header header;
-    RWKV_ASSERT_NULL_MSG(RWKV_ERROR_FILE, rwkv_fread_file_header(file, header), "invalid file header");
+    RWKV_ASSERT_NULL_MSG(RWKV_ERROR_FILE, rwkv_fread_file_header(file, header), "Fnvalid file header");
```
Comment (saharNooby) [inline rwkv.cpp:896]: `Invalid `

File: rwkv.cpp [hunk 18]
```diff
@@ -1141,15 +1141,15 @@ bool rwkv_quantize_model_file(const char * in_path, const char * out_path, const
...
+        RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_MODEL_PARAMS, rwkv_fread_tensor_header(in_file, header), "Failed to read tensor header");
+        RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_MODEL_PARAMS, rwkv_fread_string(in_file, header.key_length, name), "Failed to read tensor name");
 
         const char * name_str = name.c_str();
         RWKV_MSG("%*s - [%5" PRId32 ", %5" PRId32 "], type = %6s ", (int) max_key_length, name_str, header.width, header.height, rwkv_type_to_string[header.data_type]);
 
         data = header.data_type == TYPE_F16 ? out_buf : in_buf;
         size_t orig_size = rwkv_tensor_size(header), new_size = orig_size;
-        RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_MODEL_PARAMS, rwkv_fread_data(in_file, orig_size, data), "\nfailed to read tensor data of %s", name_str);
+        RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_MODEL_PARAMS, rwkv_fread_data(in_file, orig_size, data), "\nFailed to read tensor data of %s", name_str);
```
Comment (saharNooby) [inline rwkv.cpp:1173]: Is `\n` meant to be here? No line break in other messages

Comment (LoganDark) [reply rwkv.cpp:1173]: this is directly after a printf that does not add a trailing newline

File: rwkv.cpp [hunk 19]
```diff
@@ -189,7 +189,7 @@ enum rwkv_type {
 
 #define GGML_TYPE_UNKNOWN GGML_TYPE_COUNT
 
-static const enum ggml_type rwkv_type_to_ggml[TYPE_COUNT + 1] = {
+extern const enum ggml_type rwkv_type_to_ggml[TYPE_COUNT + 1] = {
```
Comment (saharNooby) [inline rwkv.cpp:192]: (just asking) What does `extern` do?

**Comment by saharNooby** [comment]
>[Name the state parts](https://github.com/saharNooby/rwkv.cpp/pull/74/commits/a8ebb07d2501dcb6fd544177022421660537030f)

It's great to see more features, but let's keep this PR as is to merge it more quickly

File: rwkv.cpp (hunk 19 above)
Comment (LoganDark) [reply rwkv.cpp:192]: It allows the variables to be used from FFI (which I am using to parse model files in rust by using rwkv.cpp as a library)

I am not sure if you want to allow this

**Comment by LoganDark** [comment]
> > [Name the state parts](https://github.com/saharNooby/rwkv.cpp/pull/74/commits/a8ebb07d2501dcb6fd544177022421660537030f)
> 
> It's great to see more features, but let's keep this PR as is to merge it more quickly

what do you mean merge it more quickly? you are reviewing it now, right?

File: rwkv.cpp (hunk 19 above)
Comment (saharNooby) [reply rwkv.cpp:192]: Hmm, maybe they should be in the header file then, as an official part of the API...

Comment (saharNooby) [reply rwkv.cpp:192]: Ok, we can add then to the header any time. Let's keep `extern`

**Comment by LoganDark** [comment]
Phew, what a relief~ The macOS build errors were making me anxious because I knew the fix was in here lol
